### PR TITLE
refactor rnn api

### PR DESCRIPTION
**Pull request by zhiyun**
to be consistent with attention layers.

### CODE REVIEW

File: axlearn/audio/asr_decoder.py [hunk 1]
```diff
@@ -523,24 +523,24 @@ def forward(self, inputs: Tensor) -> Tensor:
...
 
         Args:
-            inputs: An int Tensor of shape [batch_size, num_labels].
-            step_states: The step states returned by `init_step_states` or `extend_step`.
+            cached_states: A NestedTensor returned by `init_states` or `extend_step`.
+            data: An int Tensor of shape [batch_size, num_labels].
 
         Returns:
             (updated_step_states, outputs), where `outputs` is a Tensor of shape
                 [batch_size, output_dim].
```
Comment (ruomingp) [inline axlearn/audio/asr_decoder.py:544]:
```suggestion
            (updated_cache_states, outputs), where `outputs` is a Tensor of shape
                [batch_size, output_dim].
```

File: axlearn/common/rnn.py [hunk 2]
```diff
@@ -284,41 +284,41 @@ def initialize_parameters_recursively(
...
+            cached_states: A list of cached states from all layers returned by `init_states`
                 or `extend_step`.
+            data: A Tensor of shape [batch_size, input_dim], the inputs for the current step.
 
         Returns:
-            (updated_step_states, outputs), where `outputs` are usually a Tensor of shape
-            [batch_size, output_dim], and updated_step_states is a list of states from all layers.
+            (updated_step_states, outputs), where:
+            `updated_step_states` is a list of states from all layers;
+             `outputs` is a Tensor of shape [batch_size, output_dim].
```
Comment (ruomingp) [inline axlearn/common/rnn.py:308]:
```suggestion
            (updated_cache_states, outputs), where:
            `updated_cache_states` is a list of states from all layers;
             `outputs` is a Tensor of shape [batch_size, output_dim].
```

File: axlearn/common/rnn.py [hunk 3]
```diff
@@ -338,27 +338,26 @@ def output_dim(self):
...
 
         Args:
-            inputs: The inputs for the current step, often a Tensor of shape
-                [batch_size, input_dim].
-            step_states: The step states returned by `init_step_states` or `extend_step`.
+            cached_states: A NestedTensor returned by `init_states()` or `extend_step()`.
+            data: A Tensor of shape [batch_size, input_dim], the inputs for the current step.
 
         Returns:
             (updated_step_states, outputs), where `outputs` are usually a Tensor of shape
```
Comment (ruomingp) [inline axlearn/common/rnn.py:363]: Ditto.

**Review by ruomingp — COMMENTED** [review]
Thanks, Zhiyun!

File: axlearn/audio/asr_decoder.py (hunk 1 above)
Comment (zhiyun) [reply axlearn/audio/asr_decoder.py:544]: Done.

File: axlearn/common/rnn.py (hunk 2 above)
Comment (zhiyun) [reply axlearn/common/rnn.py:308]: Done.

File: axlearn/common/rnn.py (hunk 3 above)
Comment (zhiyun) [reply axlearn/common/rnn.py:363]: Done.

**Review by ruomingp — CHANGES_REQUESTED** [review]
Please see the internal PR discussion.

**Review by ruomingp — APPROVED** [review]

**Pull request merged**